Implement interior penalty discontinuous Galerkin method for temperature field.

spco opened the pull request:
Hi guys,

This is my implementation of IPDG for the temperature field, as discussed a couple of months ago. There are several TODOs around, which may need some tweaking, but I'd rather show you what I have for now so you can feedback.

A brief overview:
`parameters.use_discontinuous_temperature_discretization` is the control variable (defaults to false).
When `parameters.use_discontinuous_temperature_discretization==true`:

(1) Set temperature to use `FE_DGQ` instead of `FE_Q` in `setup_fes()`.
(2) `DoFTools::make_flux_sparsity_pattern` is called in `setup_system_preconditioner()` and `setup_system_matrix()`.
(3) Dirichlet-type boundary constraints are skipped in `compute_current_constraints()` - we impose boundary conditions weakly
(4) Face terms are calculated via `local_assemble_advection_face_terms()` at the end of `local_assemble_advection_terms()`.
(5) Face-term local matrices are added into `system_matrix`.
(6) `compute_viscosity()` returns zero (no stabilisation desired)

Changes I’ve had to make: `compute_material_model_input_values()` now takes an `FEValuesBase` object as argument to allow passing of `FEFaceValues` and `FESubfaceValues`

`Scratch::AdvectionSystem` now has a number of members to hold data from the faces, and its constructor requires a quadrature object for the faces.

`CopyData::AdvectionSystem` has vectors of local matrices to hold the internal-external, ext-int and ext-ext pairings (each element of the vector holds a matrix associated with each face or subface). A new member `local_neighbor_dof_indices` holds vectors of dof indices for each face (or subface) of the cell.



Any feedback gratefully received, particularly re whether these changes to `AdvectionSystem` are acceptable. I've tried to keep to the current style as far as I can - there's probably a lot of indenting tweaks needed, but I can do that later. I haven't done anything to attempt to optimize behaviour, e.g. minimising the number of local int_ext etc matrices by pre-calculating face/subface counts or various other parts as I don't know what returns we'd get.

I haven't included any DG parameter files - should these be included, eg convection-box-dg.prm?

Sam Cox

gassmoeller commented:
Hi Sam,
great stuff. Thanks a lot for your contribution. I can take a closer look at your code later this week, but let me start by giving you a few things you can look into immediately:
- We usually handle all the indentation stuff with a program called `astyle`, preferably something around version 2.04. Many linux distributions have it as a package, or you can compile it yourself, it is pretty small and self-contained. If you have it installed, indenting all of ASPECT's source code and include files is as simple as typing `./doc/indent` from the main directory of ASPECT. This will avoid the nasty red X next to your pull request.
- Providing some example models is nice, but what would be even better would to add some tests that prove your functionality works and acts as a reference for us to make sure nothing breaks it in the future. You can add tests by adding some (small model) prm files in the tests folder and adding a folder with the same name that contains some screen output and maybe a statistics file. This will include these models in our automated test suite. Of course the best option would be to have both: examples and tests :smile:.
- A large and interesting feature like this calls for a section in the manual ;-), we can discuss later where and how we best include that, but if you have already an idea, feel free to go ahead. The manual would also be the place to explain the examples from the point above.
- A general question: From what I have seen so far you only implement the DG method for the temperature field. We strive to keep the handling of the temperature and compositional field as similar as possible to avoid a lot of if-loops in the advection system and to not introduce unexpected differences between the two. How much work would it be to enable DG for all advection fields? If very little: Would you be willing to include that as well? If medium to a lot: Could you leave a TODO or some explanation what would be necessary?

spco commented:
Hi Rene, thanks for your feedback! I've corrected the formatting above.

On the question of DG for composition fields, are you suggesting all advection fields should use the same discretization, or that we should allow the user to pick and choose which fields are (dis)continuous?

tjhei commented:
How do you handle assembly of face terms for faces on interfaces between processors (especially in combination with hanging nodes)?

spco commented:
The layer of ghost cells provides the information on neighbouring cells owned by other processors. The issues that arise from the face-selection logic used in step-30 regarding faces on interfaces between processors are handled by modifying the face-selection logic for the case of equally-refined cells to rely on cell->id() rather than cell->level() and cell->index(). Have I missed another issue?

bangerth commented:
On 11/04/2015 03:47 AM, Sam Cox wrote:
> The layer of ghost cells provides the information on neighbouring cells owned
> by other processors. The issues that arise from the face-selection logic used
> in step-30 regarding faces on interfaces between processors are handled by
> modifying the face-selection logic for the case of equally-refined cells to
> rely on cell->id() rather than cell->level() and cell->index().

You can do it that way; or you could say that on faces between a locally owned 
and a ghost cell, the processor with the lower subdomain_id/MPI rank is 
responsible.

spco commented:
@bangerth : the documentation stated id() takes O(level) time to compute - I guess subdomain_id() is known anyway - would it be worth changing to subdomain_id() then?

bangerth commented:
Yes, that's likely more efficient. In the end you just need *some kind* of tie breaker; what exactly it is you do is not important as long as both processors agree on the procedure, so doing something cheap is better I suppose.

tjhei commented:
I am asking because the logic is slightly more complicated with adaptive refinement, where you typically have to assemble coming from the finer cells. Using ```id()``` is certainly not a great idea but works in contrast to ```index()```. It might be easier to use ```MeshWorker``` for the face assembly (or at least take a look at how the iteration is done based on the settings inside [LoopControl](https://www.dealii.org/developer/doxygen/deal.II/classMeshWorker_1_1LoopControl.html)).

spco commented:
I think the current algorithm works, under a couple of assumptions: 
* isotropic refinement (anisotropic would break it I'm sure, but is not currently implemented in ASPECT that I can find) 
* that `cell` comes from a `FilteredIterator<typename DoFHandler<dim>::active_cell_iterator>` over `IteratorFilters::LocallyOwnedCell()` (as currently happens in `assemble_advection_system()`

A rough proof:

* loop over all active, locally-owned cells `K`
* loop over every face `f` of `K`
* if `f` is at domain boundary, then we’re happy (no hanging nodes or processor boundary)
* else
  *  `L` is the `neighbor()` behind `K` and we have 6 cases: 
  * (1) If neighbor `L` is coarser and locally-owned
    * we’re happy (due to mesh regularity, `L` must be active, and face `f` is not on a processor boundary).
  * (2) If neighbor `L` is coarser and not locally-owned
    * due to mesh regularity, `L` must be active, so compute from the other side on processor 2, where we will be in case (6)
  * (3) If neighbor `L` is the same size, active and locally-owned
    * we’re happy (face `f` is not on a processor boundary).
  * (4) If neighbor `L` is the same size, active but not-locally owned. 
    *  Then because `L` is adjacent to `f`, `L` is a ghost cell, and we can find its `subdomain_id()`. So compare these and compute from the side with lower `subdomain_id` (no hanging nodes).
   * (5) If neighbor `L` is the same size, inactive but all its children that touch `f` are locally owned 
     * we’re happy (each subface is not on a processor boundary).
   * (6) `L` is the same size, inactive and some of its children that touch `f` are not locally owned.
     * For each cell `A` (child of `L`) lying adjacent to face `f` that is not locally-owned, `A` is a ghost cell so the subface between `A` and `L` is computable. All other children `B` lying adjacent to `f` are locally owned and the subface is not on a processor boundary, so treat as in (5)

Essentially, in any case where the `neighbor` (or `neighbor_child` as necessary) is not locally owned, it is a ghost cell so the face term is computable.

From what I can see, all the nasty cases being caught in LoopControl don't exist under the above assumptions, as it's been written for more general iterators and with more control over which faces are of interest.

bangerth commented:
On 11/04/2015 10:56 AM, Timo Heister wrote:
> I am asking because the logic is slightly more complicated with adaptive
> refinement, where you typically have to assemble coming from the finer cells.

The tie breaker is only necessary between cells of the same level.

spco commented:
So, my thoughts on what might need cleaning up in this code:

(1) Near-identical code in the equal-face and subface cases: Should this be cleaned up by a call to a function

    void local_assemble_advection_individual_face(const AdvectionField     &advection_field,
                                                  const typename DoFHandler<dim>::active_cell_iterator &cell,
                                     		      internal::Assembly::Scratch::AdvectionSystem<dim> &scratch,
                                		          internal::Assembly::CopyData::AdvectionSystem<dim> &data,
											      const FEFaceValuesBase<dim> face_finite_element_values)


(2) Removal of `global_field_range`, `global_max_velocity`, `global_entropy_variation` as input arguments to `local_assemble_advection_face_terms()` - these aren't used in the function.

(3) On the issue of composition fields - there shouldn’t be too much of an issue, although the only terms needed from the faces in the composition field (due to lack of diffusion) are the convection-dependent terms, so there will be a good number of `(advection_field.is_temperature ? . : . )` calls. Also, allowing only some fields to be dG while others staying continuous would require some fix to the parameters, but I don’t think that would make a huge amount of sense - presumably, all advection fields are continuous or discontinuous, so we’d only want one parameter.

I am happy to implement the composition fields and will work on that in the next day or so when I get a chance.

gassmoeller commented:
Hi Sam,
sorry for let you hanging so long. I do not know if people want to use some continuous and some discontinuous fields, I just know that I used fields for the weirdest different things, so I would like to keep the possibility. What about the following suggestion:
You could add another function to the struct `AdvectionField` in simulator.h that is called `is_discontinuous()` and decides whether any field (temperature or composition) is handled discontinuously. And we add a parameter to parameters.h (something like `List of discontinuous fields`). The parameter could be a multiple selection of all compositional fields (`Patterns::MultipleSelection(pattern_of_names)` Then is_discontinuous() could check if it is the temperature field and discontinuous (like you do right now) and if so return true, if not check if it is a compositional field, and the field is in this list and if so return true, and if none is true return false. This would keep the assembly more or less clean. Of course in assemble_face_terms you would still need to ask for `is_temperature()` to decide if you need particular terms or not.

bangerth commented:
My initial inclination would have been to either make all fields continuous or all discontinuous. But @gassmoeller is right -- people have (ab)used the flexibility of compositional fields in ways we could never have imagined, and keeping things flexible would allow to continue this tradition. If it isn't too much work, it might be interesting to leave it up to the user whether a field is cG or dG.

spco commented:
Hi guys,

sorry for leaving this hanging for such a long time - I had to go get my teeth into some theory for a couple of months!

I've fixed a couple of bugs in the original implementation, so that compositions can now be treated as continuous or discontinuous. However, I've hit an issue in the implementation of the flexible approach (allowing different fields to have cG or dG). The issue I have is that if we allow two composition fields, one cG and one dG, then this starts to break a number of assumptions in the code. For example, `Introspection` has a `struct BaseElements`, enumerating the FEs. If we have cG and dG compositional fields, they no longer share the same base element, so we need an `unsigned int continuous_compositional_fields` and `unsigned int discontinuous_compositional_fields` to replace the `unsigned int compositional_fields`. This is then going to break various other places: for example, in `mesh_refinement/composition.cc`, line 44 calls `this->get_fe().base_element(this->introspection().base_elements.compositional_fields` . This is going to need tweaking to check whether the compositional field is continuous or discontinuous. 

Similarly, in `introspection.cc`, in `setup_fes()`, the two will need treating differently, and this will affect the structure of `multiplicities` and `fes`. 

I'm asking for advice - is this the right way to go, or is there a better way?

gassmoeller commented:
Thanks for coming back to this. You are right, the individual selection seems to cause more problems than I expected. Since the individual selection is a feature that is not needed right away, would it be much work to implement your new additions in a way that allows for flexible selection of discontinuous fields, but enforce either all fields being selected as discontinuous or none, for now? 
That way we could leave the rest of the code untouched, but in case somebody wants to use mixed continuous/discontinuous elements he would only need to touch the other parts of the code, not your additions.
If that is also causing much trouble, then I would say let us just go with the all continuous / discontinuous fields for now and lets worry about the mixed case when somebody actually needs it.
Let us know if there are problems or questions.

bangerth commented:
> If that is also causing much trouble, then I would say let us just go
> with the all continuous / discontinuous fields for now and lets worry
> about the mixed case when somebody actually needs it.

I think that makes sense.

spco commented:
Ok, I think I have a (nearly) finished product! I haven't yet added any test cases. To do. Any pointers on that front would be welcomed.

Let me know what needs tidying up - otherwise, I think this is about ready to go, from my POV.

I also haven't added anything to the manual yet - from a quick run through, I think it might be worth mentioning this in §2.9 Numerical methods: Accurate discretizations, as well as §5.1.2: Parameters for certain aspects of the numerical algorithm. The two parameters `use_discontinuous_temperature_discretization` and `use_discontinuous_composition_discretization` would need adding to §5.29. I don't know whether a more substantial section is either wanted or needed? There doesn't seem to be a parallel one for the continuous case, for example. It might be worth stating explicitly somewhere that, if temperature is discontinuous, then playing around with stabilization parameters will be pointless. Just a thought.

gassmoeller commented:
/run-tests

gassmoeller commented:
The overall structure looks good, it should also be compatible to the new structure of the advection assembly (as a follow-up to #718 for the Stokes system. @bangerth: This patch needs face terms similar to #722 but for the Advection equation). 
I think you need to address a compile issue with deal.II 8.2 though, and I suspect it is because of the name change of `boundary_indicator` to `boundary_id`. For a possible fix look at assembly.cc:1237.

I do not feel competent enough to review the math of the patch, but I guess we need a benchmark and tests for that anyway.
Concerning the tests: It might be easiest to copy a few tests for the advection system and just make sure the discontinuous formulation gives (nearly) identical results to the continuous one for now. Tests that come to mind could be: diffusion.prm, time_dependent_temperature_bc_2.prm, composition_stabilization.prm.
A interesting benchmark case that is popular in the geosciences would be a variation of the test van_keken_smooth_tracer.prm with discontinuous compositional and temperature fields. It is not a perfect benchmark, but when comparing the fields at an end time of 2000 and with say global refinement 6, then it would be interesting to see how much the artificial diffusion differs to the continuous formulation.

I will leave a few minor comments in the code, but I think it might be worth merging this soon (when it compiles) to not block the restructuring of the assembly by @bangerth. Essentially so far nothing old has been changed, only the new functions are missing tests and documentation, so nobody will use it yet :smile:

gassmoeller commented on include/aspect/simulator.h:
```diff
@@ -1043,7 +1062,7 @@ namespace aspect
       void
       compute_material_model_input_values (const LinearAlgebra::BlockVector                            &input_solution,
-                                           const FEValues<dim,dim>                                     &input_finite_element_values,
+                                           const FEValuesBase<dim,dim>                                     &input_finite_element_values,
```
I think this change will be merged into the developer version with #723 soon anyway, will be removed during the rebase before merging.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -463,6 +532,9 @@ namespace aspect
            * that correspond only to the variables listed in local_dof_indices
            */
           FullMatrix<double>          local_matrix;
+          std::vector<FullMatrix<double> >         local_matrices_int_ext;
```
Maybe add a note that these are the face contributions.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                ||
+                (!advection_field.is_temperature()))
+              {
+                //We are either in the case of a Dirichlet temperature boundary, or we are assembling for a composition variable. In the former, impose the Dirichlet value weakly using a matrix term and RHS term. In the latter, if this is an inflow boundary then impose a zero Dirichlet value by assembling the corresponding matrix term but no RHS term
```
Please wrap this line. Also why is the composition case with a fixed boundary composition handled differently (maybe a naive question)? Is it because of the missing diffusion term?

spco replied:
In the composition case, Dirichlet boundary conditions are only imposed on the inflow boundary, because it is a pure advection equation. I've implemented this following the equations stated in the manual (§2.1.2), which say that the Dirichlet conditions are homogeneous. I’m aware that there are actually fixed_composition_boundary_indicators implemented, so perhaps I should change this to handle these as well. If so, I think the manual might want to be updated to reflect this. However it would probably be wise to explicitly state that Dirichlet conditions can only be imposed on the inflow boundaries, regardless of the boundary indicators passed. What do you think - should I implement this?

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                    for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+                      {
+                        scratch.face_grad_phi_field[k]      = scratch.face_finite_element_values[solution_field].gradient (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                        scratch.face_phi_field[k]      = scratch.face_finite_element_values[solution_field].value (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
```
Please align the lines or just separate everything with one space on either side of the equal sign.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                            ExcMessage ("The sum of density times c_P and the latent heat contribution "
+                                        "to the left hand side needs to be a non-negative quantity."));
+
+                    const double penalty = (advection_field.is_temperature()
```
Could you document where the magic 10 comes from? Also why is temperature and composition handled differently? Due to the thermal diffusion?

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                            * scratch.face_phi_field[i]
+                            * fixed_temperature_value
+
+                            + (current_u * scratch.face_finite_element_values.normal_vector(q) < 0
```
This implies there is a qualitative difference when material is flowing in vs. flowing out? I guess there is no space to write a lengthy explanation in a comment here, but if you can explain it in a few words, it would be nice to have a comment.

gassmoeller replied:
Also I think I remember we had a similar case with a different condition in the Stokes system, and evaluating this expression outside of the loops for matrix and rhs and using it as a constant bool inside actually saved quite some time in the assembly. I am not sure if the same applies here, but maybe it is worth the change, because then you also have more space to explain this expression.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                                            :
+                                                            0.0);
+
+                    Tensor<1,dim> current_u = scratch.face_current_velocity_values[q];
```
please make this `const`

gassmoeller replied:
ah, forget that, just saw the line below. sorry

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+            //interior face - no contribution on RHS
+            Assert (cell->neighbor(f).state() == IteratorState::valid,
+                    ExcInternalError());
+            typename DoFHandler<dim>::cell_iterator neighbor = cell->neighbor(f); //note: NOT active_cell_iterator, so this includes cells that are refined.
```
Can this be made `const`? This is mostly to avoid the need for readers (like me) to think about where you might change this afterwards.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                //loop over subfaces
+                for (unsigned int subface_no=0; subface_no<face->number_of_children(); ++subface_no)
+                  {
+                    typename DoFHandler<dim>::cell_iterator neighbor_child      //note: NOT active_cell_iterator, so this includes cells that are refined.
```
Same here. Is this `const`?

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                                   neighbor_face_heating_model_outputs);
+
+                    std::vector<types::global_dof_index> local_neighbor_dof_indices;
+                    local_neighbor_dof_indices.resize(scratch.face_finite_element_values.get_fe().dofs_per_cell);
```
You can combine the two lines above by calling the constructor `std::vector<types::global_dof_index> local_neighbor_dof_indices (scratch.face_finite_element_values.get_fe().dofs_per_cell);`
Then everyone immediately knows how large this vector is going to be.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                            scratch.face_grad_phi_field[k]      = scratch.subface_finite_element_values[solution_field].gradient (scratch.subface_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.face_phi_field[k]      = scratch.subface_finite_element_values[solution_field].value (scratch.subface_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.neighbor_face_grad_phi_field[k]      = scratch.neighbor_face_finite_element_values[solution_field].gradient (scratch.neighbor_face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.neighbor_face_phi_field[k]      = scratch.neighbor_face_finite_element_values[solution_field].value (scratch.neighbor_face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
```
Please align the equal signs above.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+
+                        const double penalty = (advection_field.is_temperature()
+                                                ?
+                                                10
```
Again the magic 10 :smirk:. Does it have a name, or is it empirical?

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                                :
+                                                0.0);
+
+                        Tensor<1,dim> current_u = scratch.face_current_velocity_values[q];
```
`const`

gassmoeller replied:
Sorry, I am too fast with this.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                    * scratch.face_phi_field[i]
+                                    * scratch.face_phi_field[j]
+
+                                    - (current_u * scratch.subface_finite_element_values.normal_vector(q) < 0
```
See comment above.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                    * scratch.neighbor_face_phi_field[j]
+                                    * scratch.face_phi_field[i]
+
+                                    + (current_u * scratch.subface_finite_element_values.normal_vector(q) < 0
```
Same here.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                    * scratch.face_phi_field[j]
+                                    * scratch.neighbor_face_phi_field[i]
+
+                                    - (!(current_u * scratch.subface_finite_element_values.normal_vector(q) < 0)
```
And here

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                      }
+                  }
+              }
+            else
```
Please add a comment line and remind me which `if` this `else` belongs to.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                                                   neighbor_face_heating_model_outputs);
+
+                    std::vector<types::global_dof_index> local_neighbor_dof_indices;
+                    local_neighbor_dof_indices.resize(scratch.face_finite_element_values.get_fe().dofs_per_cell);
```
combine lines

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                            scratch.face_grad_phi_field[k]      = scratch.face_finite_element_values[solution_field].gradient (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.face_phi_field[k]      = scratch.face_finite_element_values[solution_field].value (scratch.face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.neighbor_face_grad_phi_field[k]      = scratch.neighbor_face_finite_element_values[solution_field].gradient (scratch.neighbor_face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+                            scratch.neighbor_face_phi_field[k]      = scratch.neighbor_face_finite_element_values[solution_field].value (scratch.neighbor_face_finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
```
align equals

spco commented:
Thanks for all your comments @gassmoeller . I think I've now covered them all. It still remains to implement the fixed_composition_boundary_indicators (see comment above). Let me know if that is worth doing - I suspect so, it's only a couple of lines.

On the tests front - I'll get onto this.

And on the issue of the 'magic' 10 - this is largely empirical. It needs to be large enough that your bilinear form is coercive, but not so large that you always enforce continuity at all costs. Perhaps I should move this to an input parameter?

bangerth commented:
> In the composition case, Dirichlet boundary conditions are only imposed on the 
> inflow boundary, because it is a pure advection equation. I've implemented 
> this following the equations stated in the manual (§2.1.2), which say that the 
> Dirichlet conditions are homogeneous. I’m aware that there are actually 
> fixed_composition_boundary_indicators implemented, so perhaps I should change 
> this to handle these as well. If so, I think the manual might want to be 
> updated to reflect this. However it would probably be wise to explicitly state 
> that Dirichlet conditions can only be imposed on the inflow boundaries, 
> regardless of the boundary indicators passed. What do you think - should I 
> implement this?

Yes, that would make sense. Maybe create a testcase (or use one in the testsuite) where this feature is demonstrated using the continuous FE space, and then see that the solution looks similar with your approach?

As for the manual being out of data -- can you open a github issue that points to the page where this is the case? This way it doesn't get forgotten, and someone will get around to fix it.

bangerth commented:
On 01/22/2016 05:23 AM, Sam Cox wrote:
> And on the issue of the 'magic' 10 - this is largely empirical. It needs to be
> large enough that your bilinear form is coercive, but not so large that you
> always enforce continuity at all costs. Perhaps I should move this to an input
> parameter?

Either that (there is already a section in the input file for transport 
stabilization parameters) or at least put a sentence such as the one above 
into the code that explains where the number comes from.

spco commented:
I'll add a couple more tests or cookbooks.

On the subject of van_keken_smooth_tracer.prm, I have hit a stumbling block - I can't compile with any calls to GridTools::find_active_cell_around_point, which are necessary to run that example. I currently just comment them out as they don't touch anything I use (this only came to light when I updated my master branch a couple of weeks ago). This is probably in the same vein as #641, and/or https://github.com/dealii/dealii/issues/1958 - I get an undefined symbols error. I still use the latest Mac OSX .dmg package. I'm guessing that instead of a missing header, there's now an issue with the implicit instantiations? I'm happy to open a deal.II issue if that's the right thing, but it may be more of a .dmg issue. I know you guys love compiler issues ;)

bangerth commented:
On 01/26/2016 09:17 AM, Sam Cox wrote:
> I know you guys love compiler issues ;)

Yes, we're quite fond of them :-( Please open a github issue for deal.II 
that shows a code snippet and the complete error message you get.

spco commented:
> Yes, we're quite fond of them :-( Please open a github issue for deal.II that shows a code snippet and the complete error message you get.

Happily, I've fixed my setup, using the nuclear option. It all works fine - false alarm!

gassmoeller commented:
What is the current plan forward for this PR? I think it somewhat blocks solving #743, therefore we might just merge it and handle the documentation in a new PR. We have tests now and as long as there is no other documentation users will likely ignore the new option anyway. @spco, @bangerth what are your thoughts on this?

bangerth commented:
Good point with #743 -- I would have totally mowed over the current pull request if I had moved the code already....

I'll take a look here soon. @spco -- there are some conflicts with git right now. Can you rebase?

bangerth commented on include/aspect/simulator.h:
```diff
@@ -249,6 +249,12 @@ namespace aspect
         is_temperature () const;
 
         /**
+         * Return whether this object refers to a discontinuous field.
```
Maybe "refers to a field discretized by discontinuous finite elements"?

spco replied:
Definitely better!

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -503,6 +586,7 @@ namespace aspect
            * any other variable outside the block we are currently considering)
            */
           std::vector<types::global_dof_index>   local_dof_indices;
+          std::vector<std::vector<types::global_dof_index> >   local_neighbor_dof_indices;
```
Let's just call it `neighbor_dof_indices` (to me, "local" always refers to the "current cell") :-)

spco replied:
Sure

bangerth commented:
As a general question (@tjhei and others, please chime in):

In the current form of the patch, you initialize a lot of fields in the scratch and copy objects that only pertain to face integrals. That's going to cost time (and memory, but I'm not so concerned about this) that's only really necessary if you use the DG discretization. Would there be a way to do this only if you do use the DG element, e.g., as a `bool` constructor argument to these structs, and initialize them to zero size otherwise?

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -804,6 +913,14 @@ namespace aspect
                                               +
                                               (parameters.stokes_velocity_degree+1)/2),
+                                  //TODO: give this a 1-point QGauss to save space, as this is never going to be used here? (discontinuous temperature field doesn't require stabilization)
+                                  QGauss<dim-1>((advection_field.is_temperature()
```
I don't think I quite understand the comment. Can you elaborate (and or address the TODO)?

spco replied:
If the advection_field is discontinuous, we're never going to want to add stabilization (not that I've ever come across at least - perhaps some do!). Thus when we call compute_viscosity below, we are always going to be in the context of a continuous field. These don't need a face _quadrature in the scratch object, so it's probably better to minimise the memory and computation in the scratch constructor by passing a minimal quadrature rule. Is that the best way to proceed?

bangerth replied:
Thinking about this some more, you can create an invalid quadrature formula using `Quadrature<dim-1>()`. I think this is exactly what you want.

bangerth replied:
In other words do something like
```
(advection_field.is_temperature()
 ?
 QGauss<dim-1>(parameters.temperature_degree)
 :
 Quadrature<dim-1>())
```

bangerth replied:
But I also have another question: you only treat the case of the temperature equation here, but below you also assemble things for the composition equation. What did we decide: that the DG method is to be used only for the temperature, or for all advected quantities? In either case, either the code here or the assembly below seems wrong.

spco replied:
Fixed to only assemble face terms if advection_field is discontinuous (and never in `get_artificial_viscosity`) in e4844ff.

spco commented:
Right, I think everything is finally up to date on this branch. The only things remaining are the general discussion about initializing some fields in a minimal way, and the TODO about initializing the scratch object with a minimal quadrature. What are your thoughts on these both?

bangerth commented:
I think we really ought to find a way to initialize only those fields we actually need. For the minimal quadrature, I'm ok with whatever approach you choose as long as the comment in front of it explains what you're doing.

I'm sorry if I gave the impression that I'm done with the review -- I simply ran out of time, but I'm sure I will find numerous other things I'd like to ask you to address, if that's ok with you. I hope to get to it this week.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+    // the system we are currently trying to assemble
+    const unsigned int advection_dofs_per_cell = data.local_dof_indices.size();
+
+    Assert (advection_field.is_discontinuous(introspection), ExcMessage("Face terms only need assembling in the case of discontinuous temperature or compositional discretization."));
```
break the statement onto 2 lines (e.g., at `ExcMessage`) to roughly fit it into 80 character lines

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+         ?
+         introspection.component_indices.temperature
+         :
+         introspection.component_indices.compositional_fields[advection_field.compositional_variable]
```
I think you can get this more concisely using `advection_field.component_index(introspection)`.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+         ?
+         introspection.extractors.temperature
+         :
+         introspection.extractors.compositional_fields[advection_field.compositional_variable]
```
There is a similar issue here, but no function currently exists. I've opened #755 for this.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+          {
+            scratch.face_finite_element_values.reinit (cell, f);
+
+            scratch.face_finite_element_values[introspection.extractors.velocities].get_function_values(current_linearization_point,
```
can't you move all of these initializations into the `if` statement below that checks that it is indeed a Dirichlet-type boundary?

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+                for (unsigned int subface_no=0; subface_no<face->number_of_children(); ++subface_no)
+                  {
+                    const typename DoFHandler<dim>::cell_iterator neighbor_child      //note: NOT active_cell_iterator, so this includes cells that are refined.
+                      = cell->neighbor_child_on_subface (f, subface_no);
```
You should be able to make this an active cell iterator -- the children of an active neighbor must be active.

spco replied:
Ah yes! This over-complicates things as it handles anisotropic refinement, which we don't use. Good point.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+                  {
+                    const typename DoFHandler<dim>::cell_iterator neighbor_child      //note: NOT active_cell_iterator, so this includes cells that are refined.
+                      = cell->neighbor_child_on_subface (f, subface_no);
+                    Assert (!neighbor_child->has_children(), ExcInternalError());
```
And then you can drop this assertion because it would already be encoded in the type.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+            else
+              {
+                if (!cell->neighbor_is_coarser(f) &&
+                    (cell->subdomain_id() < neighbor->subdomain_id() ||
```
I would move this `else-if` above the `if` in line 1902 since, conceptually, it is the simpler case to treat.

Also, you know (and can test via an assertion) that `cell` is locally owned, so I think the condition here is not quite correct. You need to assemble *always* if the neighbor is a ghost cell, because you're never going to see that neighbor in `cell`. If you do this, then statements such as `cell->subdomain_id() == neighbor->subdomain_id()` may in fact be easier to state as `neighbor->is_locally_owned()`.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+
+                        const double penalty = (advection_field.is_temperature()
+                                                ?
+                                                10
```
and here, with the magic constant.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1816,6 +2808,31 @@ namespace aspect
+      {
+        const unsigned int test_output = 0;
+        for (unsigned int f=0; f<GeometryInfo<dim>::max_children_per_face
+             * GeometryInfo<dim>::faces_per_cell; ++f)
```
This loop is doing a lot of copying: In 3d, this is 24 (sub-)faces, and each matrix has size at least 116 (if no compositional fields). But in most cases, only a subset of these 24 (sub-)face matrices will have been filled. Can we somehow keep track of which ones have been filled, and only copy those? For example, by keeping a flag in the copydata structure for each matrix, that is reset to false at the beginning of the loop over the faces, and set to true everytime you write something into it?

The entries you copy are also mostly zero (namely everything in the local matrices that has to do with velocities and pressures), and we could probably avoid some copying. But that's a pre-existing issue, and so maybe not important.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1855,6 +2872,7 @@ namespace aspect
          std_cxx11::bind (&Simulator<dim>::
                           copy_local_to_global_advection_system,
                           this,
+                          advection_field,
```
make this a `std_cxx11::cref(advection_field)`.

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -96,6 +96,20 @@ namespace aspect
+      {
+        return true;
+      }
+    return false;
```
This function is a bit hard to read. How about
```
  if (field_type == temperature_field)
    return introspection.use_discontinuous_temperature_discretization;
  else if (field_type == compositional_field)
    return introspection.use_discontinuous_composition_discretization;

  Assert (false, ExcInternalError());
  return false;
```

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -599,7 +599,18 @@ namespace aspect
+                         "Whether to use a temperature discretization that is discontinuous "
+                         "as opposed to continuous. This then requires the assembly of face terms "
+                         "between cells, and weak imposition of boundary terms for the temperature "
+                         "field via the interior-penalty discontinuous Galerkin method");
```
period at the end of the sentence. Same below.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -638,6 +649,14 @@ namespace aspect
                            "different value than described there: It can be chosen as stated there for "
                            "uniformly refined meshes, but it needs to be chosen larger if the mesh has "
                            "cells that are not squares or cubes.) Units: None.");
+        prm.declare_entry ("Discontinuous penalty", "10",
```
I think you need to actually make use of this parameter in the assembly. IIRC, you're currently still hardcoding this value.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -638,6 +649,14 @@ namespace aspect
                            "cells that are not squares or cubes.) Units: None.");
+        prm.declare_entry ("Discontinuous penalty", "10",
+                           Patterns::Double (0),
+                           "The value used to penalise discontinuities in the discontinuous "
```
penalise -> penalize

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -638,6 +649,14 @@ namespace aspect
+                           Patterns::Double (0),
+                           "The value used to penalise discontinuities in the discontinuous "
+                           "Galerkin method. This is used only for the temperature field, "
+                           "as pure advection does not use the interior penalty method. This "
```
Do you want to suggest that it is *not* used for the composition field? If so, maybe state this explicitly.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -638,6 +649,14 @@ namespace aspect
+                           "Galerkin method. This is used only for the temperature field, "
+                           "as pure advection does not use the interior penalty method. This "
+                           "is largely empirically decided - it must be large enough to ensure "
+                           "the bilinear form is coercive, but not so large as to penalise "
```
penalize

in the line above, use `--` instead of `-` for a long dasg.

bangerth commented:
@spco -- I think I read through the patch once in its entirety now. Thanks a lot, this was clearly a massive amount of work!

I had many minor comments, but the only one that are really worth discussing are
* The (potentially) unnecessary initialization of a lot of fields in the scratch object. I think if you pass an empty `Quadrature<dim-1>()` object for the face quadrature, some of the fields automatically are left empty whose size will be `face_quadrature.size()`, but there are numerous others for which I think it would be very useful to pass down a flag that indicates whether these fields are needed or not.
* The (potentially) expensive copying of local contributions into the global matrix. In this case, I would try to keep a flag around during assembly that indicates whether you're going to put something into a matrix, and only copy that if the flag has been set.

Other than that just minor comments that should be easy enough to address.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1816,6 +2808,31 @@ namespace aspect
                                                     system_matrix,
                                                     system_rhs);
+
+    if (advection_field.is_discontinuous(introspection))
```
Please leave a comment somewhere around here that says that we're in the following going to copy the DG contributions element by element, rather than using `ConstraintMatrix::copy_local_to_global`, and that that is ok because there are no constraints on discontinuous fields.

spco commented:
Thanks for the comments @bangerth - the code is cleaner for it! The diffs aren't particularly useful to look at, as there is so much similar-looking code around that it picks out all the changes wrongly. Would you mind checking you agree with the logic of face assembly? Since our `cell` iterator is always locally owned and active, I take the approach that the coarser element always does the assembly, under the control of its own subdomain. In the case of equally-sized cells, faces on subdomain boundaries are handled by the cell with the lower subdomain_id; faces not on subdomain boundaries are handled by the cell with lower index. 

I haven't yet addressed the two larger issues, with (1) the face quadrature and (2) the local to global copy, I'll separate them into a new commit to keep it clean.

I think (2) is fairly simple. For (1) - are you suggesting something like 
`AdvectionSystem<dim>::
 AdvectionSystem (const FiniteElement<dim> &finite_element,
                         const FiniteElement<dim> &advection_element,
                         const Mapping<dim>       &mapping,
                         const Quadrature<dim>    &quadrature,
                         const Quadrature<dim-1>  &face_quadrature,
                         const unsigned int        n_compositional_fields,
                         const bool use_face_terms)`
 with for example `          face_phi_field (use_face_terms ? advection_element.dofs_per_cell : 0, Utilities::signaling_nan<double>()),` inside the constructor? Or is there a neater way?

bangerth commented:
On 02/16/2016 12:57 PM, Sam Cox wrote:
> I think (2) is fairly simple. For (1) - are you suggesting something like
> |AdvectionSystem<dim>::
> AdvectionSystem (const FiniteElement<dim> &finite_element,
> const FiniteElement<dim> &advection_element,
> const Mapping<dim> &mapping,
> const Quadrature<dim> &quadrature,
> const Quadrature<dim-1> &face_quadrature,
> const unsigned int n_compositional_fields,
> const bool use_face_terms)|
> with for example |face_phi_field (use_face_terms ?
> advection_element.dofs_per_cell : 0,
> Utilities::signaling_nan<double>()),| inside the constructor? Or is
> there a neater way?
>

That's one way. I think it might be simpler if in the constructor you 
don't use a separate argument `use_face_terms`, but simply asked 
`face_quadrature.size() > 0`. The two should always yield the same answer.

spco commented:
http://www.math.clemson.edu/~heister/aspect-logs/e4844ffc987bbe449ef4ab011ebbe05d43ee6447/
Looks like 8.2.0 complains about using a Quadrature() in the constructor of Scratch::AdvectionSystem, in the setup of face_finite_element_values. Any ideas what I could do to get around this? It compiles fine locally, using 8.4.pre.v4 Mac package, so I'm not entirely sure if it is just an 8.2.0 issue.

$clang --version
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.3.0
Thread model: posix

bangerth commented:
Ah, that's very annoying. @tjhei -- may we require 8.4 once that is out?

spco commented:
Local face matrices are now only copied if they were assembled. This commit will obviously still fail the tests. @bangerth - do you think it's worth also pulling a similar trick on the data structure as we are on the scratch object? `local_matrices_int_ext` etc (which is quite large), `assembled_matrices` and `neighbor_dof_indices` in `DataCopy::AdvectionSystem` are not needed unless we're using DG. Is this worth streamlining? I guess it would require passing `advection_field`, or at least the result of `advection_field.is_discontinuous(introspection)` to the data constructor?

bangerth commented:
On 02/18/2016 04:16 AM, Sam Cox wrote:
> Local face matrices are now only copied if they were assembled. This
> commit will obviously still fail the tests. @bangerth
> <https://github.com/bangerth> - do you think it's worth also pulling a
> similar trick on the data structure as we are on the scratch object?
> |local_matrices_int_ext| etc (which is quite large),
> |assembled_matrices| and |neighbor_dof_indices| in
> |DataCopy::AdvectionSystem| are not needed unless we're using DG. Is
> this worth streamlining? I guess it would require passing
> |advection_field|, or at least the result of
> |advection_field.is_discontinuous(introspection)| to the data constructor?
>

Yes. Memory allocation is expensive and we should avoid it if something 
is not needed. It also helps find bugs: having data structures that are 
initialized but not used is an invitation for bugs (e.g., someone 
writing something into it -- intentionally or by accident, expecting it 
to be copied into the global matrix, but this not happening). Much 
better to leave unused variables uninitialized (in this case to leave 
the matrix size at zero) which would trigger an assertion for every 
attempt to write something into them.

spco commented:
I think the only current outstanding issue is the invalid face quadrature.

bangerth commented:
The issue with the quadrature is in fact not that it's not possible to create a quadrature rule with no points, but that it's not possible to create a FEValuesBase object with such a quadrature formula:
```

--------------------------------------------------------
An error occurred in line <2164> of file </root/deal.II/dealii/source/fe/fe_values.cc> in function
    dealii::FEValuesBase<dim, spacedim>::FEValuesBase(unsigned int, unsigned int, dealii::UpdateFlags, const dealii::Mapping<dim, spacedim>&, const dealii::FiniteElement<dim, spacedim>&) [with int dim = 2; int spacedim = 2]
The violated condition was: 
    n_q_points > 0
The name and call sequence of the exception was:
    ExcMessage ("There is nothing useful you can do with an FEValues " "object when using a quadrature formula with zero " "quadrature points!")
Additional Information: 
There is nothing useful you can do with an FEValues object when using a quadrature formula with zero quadrature points!
```
I think that's actually quite reasonable.

What do you think of the following proposal: While the scratch objects have member variables for the `FEValues` object, they could have a `std_cxx11::shared_ptr<FEFaceValues>` member for the face integration. If the face quadrature formula passed down has no quadrature points, then that pointer would simply be initialized with NULL. Otherwise, you would create an object (via `operator new`) initialized with an appropriate `FEFaceValues` object. This way you're not just creating an invalid `FEFaceValues` object, you're in fact not creating one at all.

This should avoid the error above.

tjhei commented:
>  may we require 8.4 once that is out?

I would be happy to require 8.4.0 after the next ASPECT release but I would prefer the next ASPECT release (soon!?) to be at 8.3 or 8.2.1.

I think using a ``shared_ptr`` for some of the objects that are unused is cleaner than constructing an invalid ``FEFaceValues``.

spco commented:
I've implemented the shared_ptr method. Let me know what you think! It all seems to work locally - just waiting for the tests to run here.

Re the errors for 8.2.0: It also didn't work in 8.4.0 either, I just hadn't tested it enough locally :)

bangerth commented:
OK, great. Let me look through it one more time. In the meantime, would you mind squashing the 27 commits into one?

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -301,12 +360,32 @@ namespace aspect
+
+          local_dof_indices (finite_element.dofs_per_cell),
+          neighbor_dof_indices ((field_is_discontinuous ? GeometryInfo<dim>::max_children_per_face
+                                 * GeometryInfo<dim>::faces_per_cell : 0),
```
Please break these lines (starting at line 368) differently so that an expression in either of the branches does not unnecessarily crosses lines. Maybe like
```
  field_is_discontinuous
  ?
  ...
  :
  0
```
?

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -586,6 +676,8 @@ namespace aspect
 
     viscosity_per_cell = 0.0;
 
+    if (advection_field.is_discontinuous(introspection))
```
Add the same comment from above:
```
  //  discontinuous Galerkin doesn't require an artificial viscosity
```

bangerth commented:
I had only two really minor issues left. Please address them, squash everything, and then this is ready to merge!

I really appreciate all of the work you've put into this, and your patience with both my slowness in getting this reviewed and with addressing all of the comments I've had. Many thanks, @spco !

tjhei commented on include/aspect/assembly.h:
```diff
@@ -240,8 +265,18 @@ namespace aspect
            * that correspond only to the variables listed in local_dof_indices
            */
           FullMatrix<double>          local_matrix;
+          std::vector<FullMatrix<double> >         local_matrices_int_ext;
```
can you add a comment what these matrices are? What is the length of the vector?

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -1551,6 +1648,898 @@ namespace aspect
+                                                         cell,
+                                                         true,
+                                                         scratch.face_material_model_inputs);
+                    material_model->evaluate(scratch.face_material_model_inputs,
```
here (and many other places): do we need to average the material model outputs?

tjhei commented on tests/discontinuous_composition_1/screen-output:
```diff
@@ -0,0 +1,456 @@
+
+
++---------------------------------------------+------------+------------+
+| Total wallclock time elapsed since start    |       105s |            |
```
do you think that test requires 32 timesteps to be useful? 105 seconds is not exactly quick.

tjhei commented on include/aspect/assembly.h:
```diff
@@ -161,9 +174,14 @@ namespace aspect
+          std::vector<Tensor<1,dim> > face_mesh_velocity_values;
 
           std::vector<SymmetricTensor<2,dim> > current_strain_rates;
+          std::vector<double>         current_pressure_values;
```
are those variables used?

tjhei commented:
Sorry for only looking at it now. Looks good except my minor comments above.

spco commented:
Thanks for your comments @bangerth and @tjhei . I've addressed all the issues apart from that of averaged material_model_outputs. On that front, there are `face_material_model_outputs` and `neighbor_face_material_model_outputs`. Most terms then use whichever is appropriate. Some instead (eg penalty term and density_c_P_and_latent_heat) use the maximum over the 2 cells, as that is more appropriate. Hopefully that makes sense? I will squash these once the tests finish.

tjhei commented:
Yes, taking the maximum makes sense. :+1:

spco commented:
Right, all tests have finally passed!

bangerth commented:
I've hit the "merge" button so we can get on with the other patches that were queued up after this.

@spco -- would you mind adding an entry to `doc/modules/changes.h` regarding the DG discretization that describes the functionality you implemented in one paragraph (as a follow-up patch)?

@tjhei -- what was your take on averaging? We can do this in a follow-up patch without having to hold this further up.

@jdannberg and @gassmoeller -- if you want, go ahead and also put the advection assemblers into the signals format.

tjhei commented:
> what was your take on averaging?

The correct thing to do would be to average, even if the assembly here takes the maximum of the values in a single quadrature point.

> @jdannberg and @gassmoeller -- if you want, go ahead and also put the advection assemblers into the signals format.

Can you take a look at https://github.com/geodynamics/aspect/pull/751 before we continue with modifying the other assemblers?

bangerth commented:
Some related discussion is in #808.
